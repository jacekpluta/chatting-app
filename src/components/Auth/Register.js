import React, { Component } from "react";
import {
  Grid,
  Form,
  Segment,
  Button,
  Header,
  Message,
  Icon,
  GridColumn
} from "semantic-ui-react";
import md5 from "md5";
import firebase from "../Firebase";

import { Link } from "react-router-dom";

class Register extends Component {
  constructor(props) {
    super(props);
    this.state = {
      username: "",
      email: "",
      password: "",
      passwordConfirmation: "",
      error: "",
      loading: false,
      //state of users in database
      usersRef: firebase.database().ref("users")
    };
  }

  //checking if the form is valid or not then returns true or false
  isFormIsValid = () => {
    let error;
    if (this.isFormEmpty()) {
      error = "Fill all fields";
      this.setState({ error: error });
      return false;
    } else if (this.isPasswordValid()) {
      error = "Password is invalid";
      this.setState({ error: error });
      return false;
    } else {
      this.setState({ error: "" });
      return true;
    }
  };

  //checking if the form is empty or not then returns true or false
  isFormEmpty = () => {
    return (
      !this.state.username.length ||
      !this.state.email.length ||
      !this.state.password.length ||
      !this.state.passwordConfirmation.length
    );
  };

  //checking if the password is valid or not then returns true or false
  isPasswordValid = () => {
    if (
      this.state.password.length < 6 ||
      this.state.passwordConfirmation.length < 6
    ) {
      return true;
    } else if (
      this.state.password.length !== this.state.passwordConfirmation.length
    ) {
      return true;
    } else {
      return false;
    }
  };

  //sets current Form Input row into a state
  handleChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  //runs on submit form button click, sets loading, updates user with display name and random photo URL from gravatar
  handleSubmit = event => {
    event.preventDefault();
    if (this.isFormIsValid()) {
      this.setState({ loading: true });
      firebase
        .auth()
        .createUserWithEmailAndPassword(this.state.email, this.state.password)
        .then(createdUser => {
          console.log(createdUser);
          createdUser.user
            .updateProfile({
              displayName: this.state.username,
              photoURL: `https://www.gravatar.com/avatar/${md5(
                createdUser.user.email
              )}?d=identicon`
            })
            .then(() => {
              this.saveUser(createdUser).then(() => {
                console.log("user saved");
              });
            })
            .then(() => {
              this.setState({ loading: false, error: "" });
            })
            .catch(err => {
              console.log("create user error", err);
              this.setState({ loading: false, error: err.message });
            });
        })
        .catch(err => {
          console.log("create user error", err);
          this.setState({ loading: false, error: err.message });
        });
    }
  };

  //saves user into the database using "usersRef" from state
  saveUser = createdUser => {
    return this.state.usersRef.child(createdUser.user.uid).set({
      name: createdUser.user.displayName,
      avatar: createdUser.user.photoURL,
      email: createdUser.user.email
    });
  };

  render() {
    const { username, email, password, passwordConfirmation } = this.state;
    return (
      <Grid textAlign="center" verticalAlign="middle" className="app">
        <GridColumn style={{ maxWidth: 450 }}>
          <Header as="h1" icon color="blue" textAlign="center">
            <Icon name="rocket" color="blue" />
            Register
          </Header>
          <Form size="large" onSubmit={this.handleSubmit}>
            <Segment stacked>
              <Form.Input
                name="username"
                icon="user"
                iconPosition="left"
                placeholder="Username"
                type="text"
                onChange={this.handleChange}
                value={username}
              />
              <Form.Input
                name="email"
                icon="mail"
                iconPosition="left"
                placeholder="Email Adress"
                type="email"
                onChange={this.handleChange}
                value={email}
                className={
                  this.state.error.toLowerCase().includes("email")
                    ? "error"
                    : ""
                }
              />

              <Form.Input
                name="password"
                icon="lock"
                iconPosition="left"
                placeholder="Password"
                type="password"
                onChange={this.handleChange}
                value={password}
                className={
                  this.state.error.toLowerCase().includes("password")
                    ? "error"
                    : ""
                }
              />

              <Form.Input
                name="passwordConfirmation"
                icon="repeat"
                iconPosition="left"
                placeholder="Password Confirmation"
                type="password"
                onChange={this.handleChange}
                value={passwordConfirmation}
                className={
                  this.state.error.toLowerCase().includes("password")
                    ? "error"
                    : ""
                }
              />
              <Button
                disabled={this.state.loading}
                className={this.state.loading ? "loading" : ""}
                color="blue"
                fluid
                size="large"
              >
                Submit
              </Button>
            </Segment>
          </Form>

          {this.state.error !== "" && (
            <Message error>
              {" "}
              <h3>Error</h3>
              {this.state.error}
            </Message>
          )}
          <Message>
            Already a user? <Link to="/login">Login</Link>
          </Message>
        </GridColumn>
      </Grid>
    );
  }
}

export default Register;
