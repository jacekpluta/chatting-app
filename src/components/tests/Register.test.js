import React from 'react';

import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Register from '../Auth/Register';

Enzyme.configure({
	adapter: new Adapter()
});

let wrapper;

beforeEach(() => {
	wrapper = mount(
		<BrowserRouter>
			<Register />
		</BrowserRouter>
	);
});

afterEach(() => {
	wrapper.unmount();
});

it('should find one button with text "Submit"', () => {
	expect(wrapper.find('button').length).toEqual(1);
	expect(wrapper.find('button').text()).toBe('Submit');
});

it('should find four input fields', () => {
	expect(wrapper.find('input').length).toEqual(4);
});

it('should find one h1 field', () => {
	expect(wrapper.find('h1').length).toEqual(1);
	expect(wrapper.find('h1').text()).toBe('Register');
});

it('should find one form field', () => {
	expect(wrapper.find('form').length).toEqual(1);
});

it('should find inputs with their correct names', () => {
	expect(wrapper.find('input').at(0).prop('name')).toEqual('username');
	expect(wrapper.find('input').at(1).prop('name')).toEqual('email');
	expect(wrapper.find('input').at(2).prop('name')).toEqual('password');
	expect(wrapper.find('input').at(3).prop('name')).toEqual('passwordConfirmation');
});

it('should find input fields with empty value', () => {
	expect(wrapper.find('input').at(0).prop('value')).toEqual('');
	expect(wrapper.find('input').at(1).prop('value')).toEqual('');
	expect(wrapper.find('input').at(2).prop('value')).toEqual('');
	expect(wrapper.find('input').at(3).prop('value')).toEqual('');
});

it('has a username text area that users can type in', () => {
	wrapper.find('input').at(0).simulate('change', { target: { name: 'username', value: 'test' } });
	wrapper.update();
	expect(wrapper.find('input').at(0).prop('value')).toEqual('test');
});

it('has a email text area that users can type in', () => {
	wrapper.find('input').at(1).simulate('change', { target: { name: 'email', value: 'test@o2.pl' } });
	wrapper.update();
	expect(wrapper.find('input').at(1).prop('value')).toEqual('test@o2.pl');
});

it('has a password text area that users can type in', () => {
	wrapper.find('input').at(1).simulate('change', { target: { name: 'password', value: 'test123' } });
	wrapper.update();
	expect(wrapper.find('input').at(2).prop('value')).toEqual('test123');
});

it('has a passwordConfirmation text area that users can type in', () => {
	wrapper.find('input').at(1).simulate('change', { target: { name: 'passwordConfirmation', value: 'test123' } });
	wrapper.update();
	expect(wrapper.find('input').at(3).prop('value')).toEqual('test123');
});

it('renders correctly', () => {
	const tree = renderer
		.create(
			<BrowserRouter>
				<Register />
			</BrowserRouter>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
