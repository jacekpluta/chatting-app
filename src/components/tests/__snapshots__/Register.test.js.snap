// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ui center aligned middle aligned grid app"
>
  <div
    className="column"
    style={
      Object {
        "maxWidth": 450,
      }
    }
  >
    <h1
      className="ui blue icon center aligned header"
    >
      <i
        aria-hidden="true"
        className="blue rocket icon"
      />
      Register
    </h1>
    <form
      className="ui large form"
      onSubmit={[Function]}
    >
      <div
        className="ui stacked segment"
      >
        <div
          className="field"
        >
          <div
            className="ui left icon input"
          >
            <input
              name="username"
              onChange={[Function]}
              placeholder="Username"
              type="text"
              value=""
            />
            <i
              aria-hidden="true"
              className="user icon"
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui left icon input"
          >
            <input
              name="email"
              onChange={[Function]}
              placeholder="Email Adress"
              type="email"
              value=""
            />
            <i
              aria-hidden="true"
              className="mail icon"
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui left icon input"
          >
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
            <i
              aria-hidden="true"
              className="lock icon"
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui left icon input"
          >
            <input
              name="passwordConfirmation"
              onChange={[Function]}
              placeholder="Password Confirmation"
              type="password"
              value=""
            />
            <i
              aria-hidden="true"
              className="repeat icon"
            />
          </div>
        </div>
        <button
          className="ui blue large fluid button"
          onClick={[Function]}
          role="button"
        >
          Submit
        </button>
      </div>
    </form>
    <div
      className="ui message"
    >
      Already a user? 
      <a
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
    </div>
  </div>
</div>
`;
